### DISCUSSION TITLE
How to bulk edit sensor properties for Configs

### DISCUSSION THREAD

**Discussion by jbatterman** [opendcs/opendcs#679] [General]
@ktarbet @hoffmanjs @DanielTOsborne @MikeNeilson 

Wondering if there's a way to write a script or edit the database tables directly to add or edit sensor properties in Config files in bulk.  For example, in the NWDM database there are separate Config files for each DCP (over 800 DCPs are currently in our database).  We need to be able to assign a value to the following properties for ALL sensors in all of those configs:  CwmsDuration, CwmsParamType, and CwmsVersion.  They will all be the same thing thankfully.  Configs can be called by "MRR-*", "NWK-*", and "NWO-*" in our case.

**Comment by ktarbet** [comment]
@jbatterman I'll do some research on this.

**Comment by ktarbet** [comment]
@jbatterman  

The SQL to insert config sensor properties will look like this, for a config with two sensors:
```
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (473020,1,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (473020,2,'CwmsDuration','0');

```
This is inserting a 'CwmsDuration' of zero '0' as an example for both sensors.
I used the 'Oracle Sql Developer' tool too do this research.

Here is a general outline how you can build that SQL statement automatically. 

First check what your office code is:

```sql
select office_code,office_id from cwms_office where office_id = 'NWDM'
```

| OFFICE_CODE | OFFICE_ID |
|-------------|-----------|
| 30          | NWDM      |

Based on office code = 30.  Let's get all configs and sensors with a config name like  NWO-%   (The percent '%' is a wildcard)

```sql

SELECT pc.id,pc.name, pc.description, 
       s.configid, s.sensornumber, s.sensorname
             
     from ccp.platformconfig pc 
     join ccp.configsensor s on s.configid = pc.id
     where 
     pc.NAME like 'NWO-%' 
     and db_office_code =30 
     order by pc.id, s.sensornumber

```
| ID   | NAME      | DESCRIPTION                                | CONFIGID | SENSORNUMBER | SENSORNAME |
|------|-----------|--------------------------------------------|----------|--------------|------------|
| 20   | NWO-GRND  | James R nr Grand Rapids, ND                | 20       | 1            | STAGE      |
| 1020 | NWO-CHMT  | Marias R nr Chester, MT                    | 1020     | 1            | Stage      |
| 1020 | NWO-CHMT  | Marias R nr Chester, MT                    | 1020     | 2            | Temp_Water |
| 1020 | NWO-CHMT  | Marias R nr Chester, MT                    | 1020     | 3            | BattVolt   |
| 2020 | NWO-HSMT  | Missouri River bl Hauser Dam nr Helena, MT | 2020     | 1            | Stage      |
| 2020 | NWO-HSMT  | Missouri River bl Hauser Dam nr Helena, MT | 2020     | 2            | Battvolt   |
| 4020 | NWO-CBCO  | South Platte R ab Cheesman, CO             | 4020     | 1            | stage      |
| 4020 | NWO-CBCO  | South Platte R ab Cheesman, CO             | 4020     | 2            | precip     |
| 4020 | NWO-CBCO  | South Platte R ab Cheesman, CO             | 4020     | 3            | battvolt   |
| 7020 | NWO-ABSD  | Foot Cr nr Aberdeen, SD                    | 7020     | 8            | Precip     |
| 7020 | NWO-ABSD  | Foot Cr nr Aberdeen, SD                    | 7020     | 9            | Stage      |
| 7020 | NWO-ABSD  | Foot Cr nr Aberdeen, SD                    | 7020     | 10           | BattVolt   |
| 7020 | NWO-ABSD  | Foot Cr nr Aberdeen, SD                    | 7020     | 12           | Precip     |
| 7020 | NWO-ABSD  | Foot Cr nr Aberdeen, SD                    | 7020     | 13           | Stage      |
| 7020 | NWO-ABSD  | Foot Cr nr Aberdeen, SD                    | 7020     | 14           | BattVolt   |


This SQL command below is similar to the one above, but it generates insert statements to add data.
This will generate the same data but as a SQL command that can be sent  back to Oracle.
```sql

SELECT  'insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) '
                           || 'values ('
                           || to_char(configid) || ','
                           || to_char(s.sensornumber) ||',' 
                           || '''CwmsDuration''' ||',' 
                           || '''0''' || ');' cmd 
     from ccp.platformconfig pc 
     join ccp.configsensor s on s.configid = pc.id
     where 
     pc.NAME like 'NWO-%' 
     and db_office_code =30 
     order by pc.id, s.sensornumber

```


```sql
CMD                                                                                                                                                                                             
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (20,1,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (1020,1,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (1020,2,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (1020,3,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (2020,1,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (2020,2,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (4020,1,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (4020,2,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (4020,3,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (7020,8,'CwmsDuration','0');
insert into ccp.configsensorproperty (CONFIGID,SENSORNUMBER,PROP_NAME,PROP_VALUE) values (7020,9,'CwmsDuration','0');

```


Start small with adding single property to a single sensor, the build on that.   If you try to insert the same data twice you will get an error.

**Comment by jbatterman** [comment]
@ktarbet thank you for looking into this!  I'll work with @hoffmanjs and John Rogers to add what we need to our system.  This looks like exactly what we need to prep our system for direct store through the routing scheduler.

**Reply by jbatterman** [reply]
@ktarbet this worked like a charm!  We ended up with over 10,000 insert statements and successfully added the necessary properties for direct store to all our GOES configs.  I appreciate your work on this, it saved a lot of time!
@hoffmanjs
